File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/login/NextCloudLoginFragment.java
package app.editors.manager.ui.fragments.login;

import android.annotation.SuppressLint;
import android.app.ActionBar;
import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.net.http.SslError;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.JsResult;
import android.webkit.SslErrorHandler;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

import app.editors.manager.R;
import app.editors.manager.app.App;
import app.editors.manager.app.WebDavApi;
import app.editors.manager.managers.tools.AccountSqlTool;
import app.editors.manager.managers.tools.PreferenceTool;
import app.editors.manager.mvp.models.account.AccountsSqlData;
import app.editors.manager.ui.activities.main.MainActivity;
import app.editors.manager.ui.fragments.base.BaseAppFragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import lib.toolkit.base.managers.utils.NetworkUtils;
import lib.toolkit.base.managers.utils.StringUtils;

public class NextCloudLoginFragment extends BaseAppFragment {

    public static final String TAG = NextCloudLoginFragment.class.getSimpleName();

    private static final String KEY_PORTAL = "KEY_PORTAL";
    private static final String LOGIN_SUFFIX = "/index.php/login/flow";
    private static final String LOGIN_HEADER = "OCS-APIREQUEST";
    private static final String BACK_PATTERN_1 = "apps";
    private static final String BACK_PATTERN_2 = "files";

    public static NextCloudLoginFragment newInstance(String portal) {
        Bundle args = new Bundle();
        args.putString(KEY_PORTAL, portal);
        NextCloudLoginFragment fragment = new NextCloudLoginFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @BindView(R.id.swipeRefresh)
    SwipeRefreshLayout mSwipe;
    @BindView(R.id.webView)
    WebView mWebView;

    @Nullable
    private Unbinder mUnbinder;
    @Nullable
    private String mPortal;
    private boolean mIsClear = false;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.next_cloud_login_layout, container, false);
        mUnbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getArgs();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        NetworkUtils.clearCookies(requireContext());
        mWebView.clearCache(true);
        mWebView.setWebChromeClient(null);
        mWebView.setWebViewClient(null);
        if (mUnbinder != null) {
            mUnbinder.unbind();
        }
    }

    @Override
    public boolean onBackPressed() {
        if (mWebView.canGoBack()) {
            mWebView.goBack();
            return true;
        } else {
            requireActivity().setResult(Activity.RESULT_CANCELED);
            requireActivity().finish();
            return super.onBackPressed();
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private void getArgs() {
        final Bundle args = getArguments();
        if (args != null && args.containsKey(KEY_PORTAL)) {
            mPortal = args.getString(KEY_PORTAL);
        }

        mWebView.getSettings().setJavaScriptEnabled(true);
        mWebView.getSettings().setUserAgentString(getString(R.string.app_name_full));
        mWebView.setWebViewClient(new WebViewCallbacks());
        mWebView.setWebChromeClient(new WebViewChromeClient());
        mWebView.clearHistory();
        mWebView.clearCache(true);

        mWebView.loadUrl(mPortal + LOGIN_SUFFIX, getHeaders());
    }

    private Map<String, String> getHeaders() {
        Map<String, String> headers = new HashMap<>();
        headers.put(LOGIN_HEADER, "true");
        headers.put("USER_AGENT", getString(R.string.app_name));
        headers.put("ACCEPT_LANGUAGE", App.getLocale());
        return headers;
    }

    private class WebViewCallbacks extends WebViewClient {

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            Uri uri = request.getUrl();
            if (uri != null) {
                if (uri.toString().contains(BACK_PATTERN_1) || uri.toString().contains(BACK_PATTERN_2)) {
                    mIsClear = true;
                    NetworkUtils.clearCookies(requireContext());
                    mWebView.clearHistory();
                    mWebView.clearCache(true);
                    mWebView.loadUrl(mPortal + LOGIN_SUFFIX, getHeaders());
                    return true;
                }

                if (uri.getScheme() != null
                        && uri.getScheme().equals("nc")
                        && uri.getHost() != null
                        && uri.getHost().equals("login")) {
                    final String path = uri.getPath();
                    if (path != null) {
                        saveUser(path);
                        return true;
                    }
                }
            }

            return super.shouldOverrideUrlLoading(view, request);
        }

        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            if (url.contains("nc") && url.contains("login")) {
                saveUser(url.substring(url.indexOf(":") + 1));
            } else {
                mSwipe.setRefreshing(true);
            }
        }

        @Override
        public void onPageFinished(WebView view, String url) {
            if (mIsClear) {
                mWebView.clearHistory();
                mIsClear = false;
            }
            mSwipe.setRefreshing(false);
        }

        @Override
        public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
            super.onReceivedError(view, request, error);
        }

        @Override
        public void onReceivedHttpError(WebView view, WebResourceRequest request, WebResourceResponse errorResponse) {
            super.onReceivedHttpError(view, request, errorResponse);
        }

        @Override
        public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
            super.onReceivedSslError(view, handler, error);
        }
    }

    private void saveUser(String path) {
        final String[] args = path.split("&");
        if (args.length == 3) {
            final String portal = args[0].substring(args[0].indexOf(":") + 1);
            final String login = args[1].substring(args[1].indexOf(":") + 1);
            final String password = args[2].substring(args[2].indexOf(":") + 1);
            AccountsSqlData account = new AccountsSqlData();
            account.setWebDav(true);
            try {
                URL url = new URL(portal);

                final StringBuilder builder = new StringBuilder()
                        .append(url.getHost());

                if (url.getPath() != null && !url.getPath().isEmpty()) {
                    builder.append(url.getPath());
                    if (url.getPort() == -1) {
                        builder.append("/");
                    }
                }

                if (url.getPort() != -1) {
                    builder.append(":").append(url.getPort());
                }
                account.setPortal(builder.toString());
                account.setLogin(login);
                account.setPassword(password);
                account.setScheme(url.getProtocol() +"://");
                if (url.getPath() != null && !url.getPath().isEmpty()) {
                    account.setWebDavPath(url.getPath() + WebDavApi.Providers.NextCloud.getPath() + login + "/");
                } else {
                    account.setWebDavPath(WebDavApi.Providers.NextCloud.getPath() + login + "/");
                }
                account.setWebDavProvider(WebDavApi.Providers.NextCloud.name());
                setPreferences(account, password);
            } catch (MalformedURLException e) {
                e.printStackTrace();
            }
        }
    }

    private void setPreferences(AccountsSqlData account, String password) {
        PreferenceTool mPreferenceTool = App.getApp().getAppComponent().getPreference();
        mPreferenceTool.setDefault();
        mPreferenceTool.setLogin(account.getLogin());
        mPreferenceTool.setPassword(password);
        mPreferenceTool.setPortal(StringUtils.getUrlWithoutScheme(account.getPortal()));
        mPreferenceTool.setScheme(account.getScheme());
        login(account);
    }

    private void login(AccountsSqlData account) {
        AccountSqlTool accountSqlTool = App.getApp().getAppComponent().getAccountsSql();
        AccountsSqlData onlineAccount = accountSqlTool.getAccountOnline();
        if (onlineAccount != null) {
            onlineAccount.setOnline(false);
            accountSqlTool.setAccount(onlineAccount);
        }
        account.setOnline(true);
        accountSqlTool.setAccount(account);

        MainActivity.show(requireContext());
        requireActivity().setResult(Activity.RESULT_OK);
        requireActivity().finish();
    }

    private class WebViewChromeClient extends WebChromeClient {

        @Override
        public void onReceivedTitle(WebView view, String title) {
            setActionBarTitle(title);
        }

        @Override
        public void onReceivedIcon(WebView view, Bitmap icon) {
            ActionBar actionBar = requireActivity().getActionBar();
            if (actionBar != null) {
                actionBar.setIcon(new BitmapDrawable(getResources(), icon));
            }
        }

        @Override
        public boolean onJsAlert(WebView view, String url, String message, JsResult result) {
            showToast(message);
            return super.onJsAlert(view, url, message, result);
        }

    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/login/NextCloudLoginFragment.kt
package app.editors.manager.ui.fragments.login

import android.annotation.SuppressLint
import android.app.Activity
import android.graphics.Bitmap
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.JsResult
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.databinding.NextCloudLoginLayoutBinding
import app.editors.manager.ui.activities.main.MainActivity
import app.editors.manager.ui.fragments.base.BaseAppFragment
import app.editors.manager.viewModels.login.NextCloudLogin
import app.editors.manager.viewModels.login.NextCloudLoginViewModel
import kotlinx.coroutines.launch
import lib.toolkit.base.managers.utils.NetworkUtils.clearCookies
import lib.toolkit.base.managers.utils.putArgs

class NextCloudLoginFragment : BaseAppFragment() {

    companion object {
        @JvmField
        val TAG: String = NextCloudLoginFragment::class.java.simpleName

        private const val KEY_PORTAL = "KEY_PORTAL"
        private const val LOGIN_SUFFIX = "/index.php/login/flow"
        private const val LOGIN_HEADER = "OCS-APIREQUEST"
        private const val BACK_PATTERN_1 = "apps"
        private const val BACK_PATTERN_2 = "files"

        @JvmStatic
        fun newInstance(portal: String?): NextCloudLoginFragment {
            return NextCloudLoginFragment().putArgs(KEY_PORTAL to portal)
        }
    }

    private val headers: Map<String, String> by lazy {
        hashMapOf(
            LOGIN_HEADER to "true",
            "USER_AGENT" to getString(R.string.app_name),
            "ACCEPT_LANGUAGE" to App.getLocale()
        )
    }

    private val viewModel: NextCloudLoginViewModel by viewModels()

    private var viewBinding: NextCloudLoginLayoutBinding? = null

    private val portal: String? by lazy { arguments?.getString(KEY_PORTAL) }
    private var isClear = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        viewBinding = NextCloudLoginLayoutBinding.inflate(inflater, container, false)
        return viewBinding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initWebView()
        collectState()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        viewBinding = null
        clearCookies()
        viewBinding?.webView?.clearCache(true)
        viewBinding?.webView?.webChromeClient = null
    }

    override fun onBackPressed(): Boolean {
        return if (viewBinding?.webView?.canGoBack() == true) {
            viewBinding?.webView?.goBack()
            true
        } else {
            requireActivity().setResult(Activity.RESULT_CANCELED)
            requireActivity().finish()
            super.onBackPressed()
        }
    }

    private fun collectState() {
        lifecycleScope.launch {
            viewModel.state.collect { state ->
                when (state) {
                    NextCloudLogin.Error -> showSnackBar(R.string.errors_unknown_error)
                    NextCloudLogin.Success -> {
                        with(requireActivity()) {
                            setResult(Activity.RESULT_OK)
                            finish()
                            MainActivity.show(this)
                        }
                    }
                    else -> Unit
                }
            }
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun initWebView() {
        viewBinding?.webView?.apply {
            settings.javaScriptEnabled = true
            settings.userAgentString = getString(R.string.app_name_full)
            webViewClient = WebViewCallbacks()
            webChromeClient = WebViewChromeClient()
            clearHistory()
            clearCache(true)
            loadUrl(portal + LOGIN_SUFFIX, headers)
        }
    }

    private inner class WebViewCallbacks : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            val uri = request.url
            if (uri != null) {
                if (uri.scheme != null && uri.scheme == "nc" && uri.host != null && uri.host == "login") {
                    val path = uri.path
                    if (path != null) {
                        viewModel.saveUser(path)
                        return true
                    }
                }
                if (uri.toString().contains(BACK_PATTERN_1) || uri.path?.contains(BACK_PATTERN_2) == true) {
                    isClear = true
                    clearCookies()
                    viewBinding?.webView?.clearHistory()
                    viewBinding?.webView?.clearCache(true)
                    viewBinding?.webView?.loadUrl(portal + LOGIN_SUFFIX, headers)
                    return true
                }
            }
            return super.shouldOverrideUrlLoading(view, request)
        }

        override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
            if (url.contains("nc") && url.contains("login")) {
                viewModel.saveUser(url.substring(url.indexOf(":") + 1))
            } else {
                viewBinding?.swipeRefresh?.isRefreshing = true
            }
        }

        override fun onPageFinished(view: WebView, url: String) {
            if (isClear) {
                viewBinding?.webView?.clearHistory()
                isClear = false
            }
            viewBinding?.swipeRefresh?.isRefreshing = false
        }
    }

    private inner class WebViewChromeClient : WebChromeClient() {
        override fun onReceivedTitle(view: WebView, title: String) {
            setActionBarTitle(title)
        }

        override fun onJsAlert(view: WebView, url: String, message: String, result: JsResult): Boolean {
            showToast(message)
            return super.onJsAlert(view, url, message, result)
        }
    }
}